storage/namespace/dynamic/registry.go

```diff
@@ -0,0 +1,264 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package dynamic
+
+import (
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/m3db/m3cluster/kv"
+	nsproto "github.com/m3db/m3db/generated/proto/namespace"
+	"github.com/m3db/m3db/storage/namespace"
+	"github.com/m3db/m3db/storage/namespace/convert"
+	"github.com/m3db/m3x/log"
+	"github.com/m3db/m3x/watch"
+
+	"github.com/uber-go/tally"
+)
+
+var (
+	errInitTimeOut           = errors.New("timed out waiting for initial value")
+	errRegistryAlreadyClosed = errors.New("registry already closed")
+	errInvalidRegistry       = errors.New("could not parse latest value from config service")
+)
+
+type dynamicInitializer struct {
+	sync.Mutex
+	opts namespace.DynamicOptions
+	reg  namespace.Registry
+}
+
+// NewInitializer returns a dynamic namespace initializer
+func NewInitializer(opts namespace.DynamicOptions) namespace.Initializer {
+	return &dynamicInitializer{opts: opts}
+}
+
+func (i *dynamicInitializer) Init() (namespace.Registry, error) {
+	i.Lock()
+	defer i.Unlock()
+
+	if i.reg != nil {
+		return i.reg, nil
+	}
+
+	if err := i.opts.Validate(); err != nil {
+		return nil, err
+	}
+
+	reg, err := newRegistry(i.opts)
+	if err != nil {
+		return nil, err
+	}
+
+	i.reg = reg
+	return i.reg, nil
+}
+
+type dynamicRegistry struct {
+	sync.RWMutex
+	opts         namespace.DynamicOptions
+	logger       xlog.Logger
+	metrics      *dynamicRegistryMetrics
+	watchable    xwatch.Watchable
+	kvWatch      kv.ValueWatch
+	currentValue kv.Value
+	currentMap   namespace.Map
+	closed       bool
+}
+
+type dynamicRegistryMetrics struct {
+	numInvalidUpdates tally.Counter
+	currentVersion    tally.Gauge
+}
+
+func newMetrics(opts namespace.DynamicOptions) *dynamicRegistryMetrics {
```

Comment:
Perhaps return just a struct here and use as a value in the `dynamicRegistry` struct?  This is consistent with how we use it in other places, the memory locality is one less jump away.. lol.

Also minor nit: `newDynamicRegistryMetrics(...)` to keep it consistent with our other metrics struct initializers?